casper/src/test/scala/coop/rchain/casper/batch1/MultiParentCasperCommunicationSpec.scala

```diff
@@ -14,55 +15,55 @@ import org.scalatest.{FlatSpec, Inspectors, Matchers}
-//
-//        deploy2      <- ConstructDeploy.sourceDeployNowF("@2!(2)")
-//        signedBlock2 <- nodes(0).addBlock(deploy2)
-//        _            <- nodes(2).syncWith(nodes)
-//        //1 receives block2
-//        //2 receives block2; asks if who has block1
-//        //1 receives request for has block1; sends i have block1
-//        //2 receives I have block1; asks for block1
-//        //1 receives request block1; sends block1
-//        //2 receives block2; asks for block1
-//        //2 receives block1; adds both block1 and block2
-//
-//        _ <- nodes(2).contains(signedBlock1.blockHash) shouldBeF true
-//        _ <- nodes(2).contains(signedBlock2.blockHash) shouldBeF true
-//
-//        /*       this test is too restrictive in the presence of block hashes (see RCHAIN-3819).
-//                 Node #1 at this point only "knows" about block2 - which is nonetheless sufficient for recovering missing blocks.
-//                 Leaving for reference:
-// _ <- nodes.toList.traverse_ { node =>
-//              for {
-//                maybeBlock1 <- node.blockStore.get(signedBlock1.blockHash)
-//                maybeBlock2 <- node.blockStore.get(signedBlock2.blockHash)
-//              } yield {
-//                withClue(s"Assertion failed for node ${node.local} --") {
-//                  maybeBlock1 shouldBe Some(signedBlock1)
-//                  maybeBlock2 shouldBe Some(signedBlock2)
-//                }
-//              }
-//            }*/
-//      } yield ()
-//    }
-//  }
+  import coop.rchain.casper.util.GenesisBuilder._
+
+  implicit val timeEff = new LogicalTime[Effect]
+
+  val genesis = buildGenesis()
+
+  "MultiParentCasper" should "ask peers for blocks it is missing" in effectTest {
+    TestNode.networkEff(genesis, networkSize = 3).use { nodes =>
+      for {
+        deploy1 <- ConstructDeploy.sourceDeployNowF("for(_ <- @1){ Nil } | @1!(1)")
+
+        signedBlock1 <- nodes(0).addBlock(deploy1)
+        _            <- nodes(1).handleReceive()
+        _            <- nodes(2).shutoff() //nodes(2) misses this block
+
+        deploy2      <- ConstructDeploy.sourceDeployNowF("@2!(2)")
+        signedBlock2 <- nodes(0).addBlock(deploy2)
+        _            <- nodes(2).syncWith(nodes)
+        //1 receives block2
+        //2 receives block2; asks if who has block1
+        //1 receives request for has block1; sends i have block1
+        //2 receives I have block1; asks for block1
+        //1 receives request block1; sends block1
+        //2 receives block2; asks for block1
+        //2 receives block1; adds both block1 and block2
+
+        // due to asynchronous processing of blocks syncWith ends before blocks are added to DAG
+        // as this test is supposed to check that node is asking for block, just checking blockstore
+        // for presence of block is enough
+        _ <- nodes(2).blockStore.contains(signedBlock1.blockHash) shouldBeF true
```

Comment:
this is fix for this tests, ptal at comment

Reply:
This test doesn't check what it says in the title, it only checks that node 2 will eventually receive block 1.

Can we check that BlockRetriever received a request from Casper  for block 1 (to not go further to TransportLayer)?

Reply:
https://github.com/rchain/rchain/pull/3355/commits/00eb5badd57112a8e68e41498402476caf77db2a